### ISSUE TITLE
Float precision issue with mass in TLorentzVector

### ISSUE THREAD

**Issue by schaed** [scikit-hep/uproot4#311]
Hi, 

I'm using uproot version 4.0.6 with uproot_methods 0.9.2. This bug affects uproot_methods TLorentzVector. I see that this is package is deprecated, but it still appears to be getting updated. Apologies if I am reporting this in the wrong place as well reporting on a deprecated package.

I have a lot of very low mass TLorentzVector objects, and I've seen that summing them together can lead to an error from math.sqrt() from negative argument.

mag2 returns a negative value, which then results in the above error.
https://github.com/scikit-hep/uproot3-methods/blob/master/uproot3_methods/classes/TLorentzVector.py#L58

>>> import uproot_methods
>>> m=uproot_methods.TLorentzVector(x=32.07181913680199, y=-212.10064225436406, z=-126.49557185391019, t=249.0309493338232)
>>> m.mass # this will crash
>>> m.mag2 # this is -4e-12, which is before the float precision. Can some protection be added for float precision?

Best,

Doug

**Comment by jpivarski** [comment]
I actually don't know _how_ you're using uproot 4.0.6 with uproot-methods—they are formally incompatible.

You're right that there was a bug fix in uproot-methods a week ago (scikit-hep/uproot3-methods#98), but that was related to writing histograms to ROOT files and that's the one feature that Uproot 4 doesn't have yet—the only recourse is to use Uproot 3. (See #310 and https://github.com/scikit-hep/uproot4/issues/25#issuecomment-802917934).

On Lorentz vectors, the [Vector](https://github.com/scikit-hep/vector) library is almost done, so it makes more sense to finish that than patch uproot-methods. On the specific issue with negative masses, Vector is following ROOT's prescription of copying the sign through a square root: `mass-squared = -4e-12 ⇒ mass = -2e-6`. This maps the spacelike region of Minkowski space onto negative mass values (instead of imaginary mass values, but we don't want to introduce complex number types just for round-off errors).

**Comment by eduardo-rodrigues** [comment]
Hi. Note also that single floating point precision (for 4-vectors) is regularly a source of issues when calculating invariant masses, just because of the calculation involved when the energy and momentum are much larger than their difference E-|p|. Using double precision can be a good solution. Else avoid using the energy and rather use mass and momentum in single precision, switching to double precision if you need E for some calculation?

**Comment by jpivarski** [comment]
Vector is letting the component dtype be completely unconstrained, governed by the backend. But that means that it can be changed using the backend's standard idiom: NumPy arrays of vectors can be changed with `.astype(np.float64)`, Awkward Arrays of vectors can be changed with [ak.values_astype](https://awkward-array.readthedocs.io/en/latest/_auto/ak.values_astype.html), etc.

(That sort of thing shouldn't be automatically changed by some operations and not others. The `E**2 - p**2` difference is often numerically sensitive, but so are others, and some only in certain situations. Anyway, adopting ROOT's spacelike-to-negative-mass representation solves this issue: no more `NaN`s for masses near zero.)

**Comment by eduardo-rodrigues** [comment]
Agreed, @jpivarski. Also some calculations involving angles in multi-body decays can trigger numerical instabilities ... The fact that dtype is unconstrained is a great thing, of course.

I am not suggesting to change precision automaticall, BTW, no worries. Just noting that in some situations that's the way to get around problems, I mean when you're stuck ... ;-)
